zebra-network/src/peer/connection.rs

```diff
@@ -859,93 +793,76 @@ where
-    /// fail_with to terminate the entire peer connection, shrinking the number
-    /// of connected peers.
-    async fn drive_peer_request(&mut self, req: Request) {
+    /// processing the request (e.g., the service is shedding load), then we
+    /// return a peer error, which terminates the entire peer connection,
+    /// shrinking the number of connected peers.
+    async fn drive_peer_request(&mut self, req: Request) -> Result<(), PeerError> {
         trace!(?req);
         use tower::{load_shed::error::Overloaded, ServiceExt};
 
         if self.svc.ready_and().await.is_err() {
             // Treat all service readiness errors as Overloaded
             // TODO: treat `TryRecvError::Closed` in `Inbound::poll_ready` as a fatal error (#1655)
-            self.fail_with(PeerError::Overloaded);
-            return;
+            Err(PeerError::Overloaded)?
         }
 
         let rsp = match self.svc.call(req).await {
             Err(e) => {
                 if e.is::<Overloaded>() {
                     tracing::warn!("inbound service is overloaded, closing connection");
                     metrics::counter!("pool.closed.loadshed", 1);
-                    self.fail_with(PeerError::Overloaded);
+                    Err(PeerError::Overloaded)?
                 } else {
                     // We could send a reject to the remote peer, but that might cause
                     // them to disconnect, and we might be using them to sync blocks.
-                    // For similar reasons, we don't want to fail_with() here - we
-                    // only close the connection if the peer is doing something wrong.
+                    // For similar reasons, we don't return an error - we only close
+                    // the connection if the peer is doing something wrong.
                     error!(%e,
                            connection_state = ?self.state,
                            client_receiver = ?self.client_rx,
                            "error processing peer request");
                 }
-                return;
+                return Ok(());
             }
             Ok(rsp) => rsp,
         };
 
         match rsp {
             Response::Nil => { /* generic success, do nothing */ }
-            Response::Peers(addrs) => {
-                if let Err(e) = self.peer_tx.send(Message::Addr(addrs)).await {
-                    self.fail_with(e);
-                }
-            }
+            Response::Peers(addrs) => self.peer_tx.send(Message::Addr(addrs)).await?,
             Response::Transactions(transactions) => {
                 // Generate one tx message per transaction.
                 for transaction in transactions.into_iter() {
-                    if let Err(e) = self.peer_tx.send(Message::Tx(transaction)).await {
-                        self.fail_with(e);
-                    }
+                    self.peer_tx.send(Message::Tx(transaction)).await?;
```

Comment:
This early exit is probably the right thing to do, but let's check.

Reply:
In this code, a peer has asked us for a list of transactions, and we're sending them back one by one.
If the peer connection fails, we stop sending transactions.
This is the correct behaviour, because we don't send messages on failed connections.